File: leetcode3/MovePiecestoObtainaString/a.go
/* Copyright (C) 2022 by iamslash */

package main

import "fmt"

// 55ms 16.46% 6.6MB 94.04%
// two pointers
// O(N) O(1)
func canChange(s string, t string) bool {
    var i, j, n = 0, 0, len(s)
    for i < n || j < n {
        for i < n && s[i] == '_' {
            i++
        }
        for j < n && t[j] == '_' {
            j++
        }
        if i == n ||
           j == n ||
           s[i] != t[j] ||
           (s[i] == 'L' && i < j) ||
           (s[i] == 'R' && i > j) {
            break
        }
        // fmt.Println(i, j)
        i++
        j++
    }
    return i == n && j == n
}

func main() {
    fmt.Println("hello world")
}

File: leetcode3/MovePiecestoObtainaString/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2022 by iamslash

# 622ms 31.18% 15.1MB 70.66%
# two pointers
# O(N) O(1)
class Solution:
    def canChange(self, s: str, t: str) -> bool:
        i, j, n = 0, 0, len(s)
        while i < n or j < n:
            while i < n and s[i] == '_':
                i += 1
            while j < n and t[j] == '_':
                j += 1
            if i == n or j == n or s[i] != t[j] or (s[i] == 'L' and i < j) or (s[i] == 'R' and i > j):
                break                
            i += 1
            j += 1
        return i == n and j == n
        
